### PR TITLE
[Feature]: Support DDTree speculative decoding

### PR DESCRIPTION
**Pull request by zzl-7**
https://github.com/vllm-project/vllm/issues/40809
Add **DDTree** (Diffusion Draft Tree) speculative decoding, implementing the draft-tree construction algorithm from [arXiv:2604.12989](https://arxiv.org/abs/2604.12989) ("Accelerating Speculative Decoding with Block Diffusion Draft Trees", Ringel & Romano).

DDTree extends the existing DFlash proposer with a dynamic, per-request best-first draft tree instead of a flat greedy sequence. Each proposal step:

1. Runs the DFlash draft model to get per-position logits.
2. Builds an independent draft tree per request via a best-first heap (`build_ddtree_tree`), selecting the `budget` most-probable token paths.
3. Pushes a per-batch `[batch, N+1, N+1]` ancestor-only attention mask to the `TreeAttentionMetadataBuilder` so the target model verifies the full tree in a single forward pass.
4. Uses depth-based RoPE positions (siblings at the same depth share a position ID) so attention correctly treats siblings as alternatives for the same output slot.

Tree verification (`ddtree_verify`) traces the longest accepted path through each request's tree using the target model's greedy predictions.

New config: `speculative_config.method = "ddtree"` with `attention_config.backend = "TREE_ATTN"` required.

# Test Plan

**Model:** `nm-testing/dflash-qwen3-8b-speculators` (Qwen3-8B + DFlash draft)  
**Dataset:** ShareGPT (200 prompts) | **budget:** 64 | `--max-num-seqs 16` | `--max-model-len 4096` | `--enforce-eager`  
**GPU:** NVIDIA RTX 5090 Laptop (24 GiB)

| method | tok/s | req/s | acc\_len | acc\_rate |
|--------|-------|-------|----------|-----------|
| dflash | 130.4 | 0.59  | 2.318    | 0.021     |
| ddtree | 179.7 | 0.81  | 3.406    | 0.038     |

DDTree vs DFlash: acceptance\_len **+1.089** · throughput **+49.3 tok/s (+37.8%)**

> Note: budget=64 is required for meaningful tree branching. At budget=8 the tree is nearly as shallow as a flat chain and gains are modest. `--max-num-seqs 16` is needed to avoid 5090 OOM at this budget/vocab size (Qwen3 vocab ≈ 152k). Extra test with better GPU is needed 

```
.venv/bin/python benchmark_ddtree_sharegpt.py \
    --model nm-testing/dflash-qwen3-8b-speculators \
    --dataset-path /tmp/sharegpt.json \
    --trust-remote-code \
    --num-speculative-tokens 64 \
    --num-prompts 200 \
    --enforce-eager \
    --max-model-len 4096 \
    --max-num-seqs 16 \
    --methods dflash ddtree

```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Review by benchislett — COMMENTED** [review]
I thought we removed tree attention? FYI @LucasWilkinson @MatthewBonanni

**Comment by MatthewBonanni** [comment]
@benchislett correct, tree attn was removed by https://github.com/vllm-project/vllm/pull/42121

**Comment by zzl-7** [comment]
> I thought we removed tree attention? FYI @LucasWilkinson @MatthewBonanni

thanks for pointing this out, if thats the case I don't think ddtree can be supported
let me comment on original issue
thanks for taking a look

**Comment by liranringel** [comment]
An author of DDTree here. Excited to see DDTree being implemented in vLLM. Great work on the implementation and speedups!

Just to add some context that may not have been clear enough in our paper: DDTree is not tied to DFlash specifically. It can be implemented for any draft model that outputs marginal per-position distributions over a block of future tokens. This includes DFlash, but also [PARD](https://docs.vllm.ai/en/latest/features/speculative_decoding/parallel_draft_model/), [P-EAGLE](https://vllm.ai/blog/2026-03-13-p-eagle), [Medusa](https://docs.vllm.ai/en/stable/api/vllm/model_executor/models/medusa/), etc.

Since DFlash is currently one of the strongest parallel drafter families with checkpoints trained for many open-weight target models, it is definitely a good starting point to support it first. I also think the current API direction is the right future-proof approach: exposing DDTree with `"method": "ddtree"`, while allowing the actual parallel drafter to be selected independently with `"model": "<drafter>"`. This should make it easier to support other block/parallel drafters later.

**Comment by avisual** [comment]
Tested this on DGX Spark (NVIDIA GB10, sm_121a, Blackwell, 128GB unified memory, aarch64) — May 27, 2026. Ran your exact benchmark command — 100 ShareGPT prompts (scaled from 200 for time), `nm-testing/dflash-qwen3-8b-speculators` drafter + `Qwen/Qwen3-8B` target, `--num-speculative-tokens 64 --max-num-seqs 16 --max-model-len 4096 --enforce-eager`, comparing `method=dflash` vs `method=dflash_ddtree` with `ddtree_budget=64`.

| Mode | Wall clock | Aggregate tok/s | Median tok/s |
|---|---|---|---|
| dflash | 1085.0s | **19.8** | 19.6 |
| dflash_ddtree | 1082.2s | **19.8** | 19.5 |

No measurable speedup. Outputs byte-identical on the 3 prompts I spot-checked (temperature=0 greedy, expected).

Likely culprit: **`--enforce-eager` is disproportionately punishing on Spark's ARM cores.** Our absolute throughput (19.8 tok/s) is ~6–9× lower than your RTX 5090 Laptop's 130.4 tok/s — almost certainly from CPU-side kernel launch overhead dominating without CUDA graphs. Whatever tree-branching advantage DDTree provides is swamped by per-step launch cost on this hardware.

If/when the dynamic-shape issues with tree attention can be reconciled with CUDA graph capture, I'd be happy to re-run on Spark — benchmark infrastructure is in place.

Output correctness on pure-transformer Qwen3-8B looks clean — no quality regressions vs DFlash. (Separately tested on Qwen3.6-27B hybrid via AEON-7's `vllm-aeon-ultimate-ddtree:qwen36-v5-m53-experimental` container in `ddtree-full` mode and saw GDN replay corruption — `self.head = self.head.next` references on LRU code and syntax errors in bug-fix outputs — but that's their experimental wrapper, separate from this PR. The PR's pure-transformer path is solid as far as I can tell.)

**Comment by benchislett** [comment]
I do not think we are ready to support/maintain tree drafting right now. I won't close the PR, but please keep it as a draft. It will be a while before tree attention is re-introduced (if at all).

**Comment by rmagur1203** [comment]
Sharing measurements from an independent implementation, in case they help
with the "is tree drafting worth maintaining" question.

- Repo: https://github.com/rmagur1203/vllm-ddtree (branch `ddtree-multireq`)
- Findings (English): [`docs/DDTREE-FINDINGS-EN.md`](https://github.com/rmagur1203/vllm-ddtree/blob/ddtree-multireq/docs/DDTREE-FINDINGS-EN.md)
  — verdict matrix, methodology, cost accounting, everything that broke along
  the way and how it was fixed, and the parts reusable regardless of DDTree.

Hardware is 2× RTX A6000 (not 5090), and we didn't test your exact
8B + DFlash pairing, so nothing below directly contradicts your number.
Where it overlaps, it's the measurement method rather than the result.

### On @avisual's "no measurable speedup under eager"

There's a mechanism for that, and it's worse than neutral. Eager execution is
launch-bound, so its GPU idle time absorbs the tree's extra GPU work. Enabling
CUDA graphs removes that slack and the overhead becomes visible. Same config
(Qwen3.5-4B GDN hybrid + DFlash, batch 1, greedy):

| | flat chain | DDTree |
|---|---|---|
| `enforce_eager=True` | 70.9 tok/s | 70.7 — no difference |
| CUDA graph | **144.9** tok/s | **99.1 (−31.6%)** |

Both arms took **the same number of steps (34)** — the tree bought no extra
acceptance at all there, only cost. Per-layer profiling hooks push in the same
direction (~10%). We reported a win for this configuration once and withdrew
it after re-measuring this way.

So I'd suggest re-running the +37.8% with `enforce_eager=False`, no
instrumentation, and counting `execute_model` calls directly so that
acceptance gain and per-step cost are separated. Also worth checking against
vLLM's *default* sampling: with `--generation-config` unset it takes
`temperature/top_k/top_p` from the model's `generation_config.json`
(1.0 / 20 / 0.95 for Qwen3.8), and a `temperature=0` benchmark doesn't
represent that traffic.

### What we tested, and where each one landed

Every drafter type we had — none, n-gram prompt lookup, block-diffusion
(DFlash / DFlash2), autoregressive draft head (EAGLE3) — on targets from 0.6B
to 27B, pure-attention and GDN-hybrid, batch 1–16, greedy and sampled.
DDTree vs. *the same drafter running a flat chain at the same budget*:

| target | drafter | batch | vs chain | what decided it |
|---|---|---|---|---|
| Qwen3-0.6B (pure attn) | n-gram | 1 | 0.95x | fewer steps (184 vs 193) but per-step cost not covered; on a 2.7 ms/step model even the chain loses to no-spec |
| Qwen3.5-4B (GDN hybrid) | DFlash | 1 | 0.68x | same 34 steps in both arms |
| Qwen3-8B (pure attn) | n-gram | 1 | ~1.01x (tie) | branching helps (+5.8%) but n-gram fires on only **28% of steps** |
| Qwen3-8B (pure attn) | EAGLE3 | 1 | **1.00x (tie)** | +17.6% acceptance on open-ended prompts, exactly cancelled by tree cost |
| Qwen3-8B | EAGLE3 | 2 / 4 / 16 | 0.98x / 0.97x / 0.96x | verification width stops being free; optimal budget 16 → 8 → 4 |
| Qwen3.8-27B-AWQ (GDN hybrid) | DFlash2, T=0 | 1 | 0.78x | structural GPU-idle bubble, below |
| same | DFlash2, default sampling | 1 | 0.69x | siblings steal spine slots; spine weaker at T=1 |
| same | DFlash2, wider trees (budget 8–32) | 1 | 0.61–0.73x | acceptance saturates at two siblings, below |

(MTP head was evaluated only as a chain baseline when choosing the production
drafter — DFlash2 dynamic-k beat it by +39.9% at concurrency 1 and tied from
8 up — never with DDTree.)

Two things the matrix says on its own. Speculation only pays on a big enough
target (0.6B: nothing beats plain decoding; 27B: chain is 2.07× base), so the
chain DDTree has to beat is strongest exactly where production runs. And the
two "bad" drafters fail for opposite reasons: n-gram is *absent* (72% of
steps have no draft to branch), DFlash2 is *too right* (rank-0 already
correct). The tree wants a drafter that is present, often wrong at rank 0,
and right at rank 1–2. EAGLE3 on open-ended prompts is the closest we found,
and it ties.

### The structural cost is GPU idle, not GPU work

Torch profiler over both arms (27B, 24 tokens, ~7.5 steps):

| | total CUDA | total CPU | wall | GPU busy |
|---|---|---|---|---|
| chain | 329.3 ms | 289.2 ms | 325 ms | **101%** |
| DDTree | 331.9 ms | 319.6 ms | 369 ms | **90%** |

GPU work differs by +2.6 ms; CPU by +30.4 ms. Chain speculative decoding
finishes verification entirely on the GPU — the rejection sampler is a kernel,
the CPU never intervenes — so it pipelines. A tree comes back to the CPU every
step to walk `child_maps`, and that round trip is a **4.9 ms/step bubble** at
27B. The biggest profiler entry is `cudaMemcpyAsync`, but that's the CPU
waiting on the stream, not transfer — pinned host buffers changed nothing.
Moving the walk to the GPU has a measured ceiling of 0.88x, so it still loses.

Of the original 20 ms/step gap, about half was our own integration overhead
and is gone; the numbers above are after that.

### How much acceptance is available from branching at all

We measured the ceiling rather than estimating it. A node's sampled token is
"what the target picked given that node's ancestor context", which stays valid
in **any sub-tree containing that node** — so one wide tree run with tracing
gives acceptance for every sub-shape offline. We also log the drafter's top-16
per depth, so sibling hit rates come out unbiased.

27B + DFlash2, 6 open-ended prompts × 192 tokens, batch 1, greedy:

| shape | tokens/step | steps |
|---|---|---|
| chain, budget 7 | 3.200 | 360 |
| spine + 1 sibling | 3.200 | 360 |
| **spine + 2 siblings** | **3.329** | **346** |
| spine + 5 / 9 / 25 siblings | 3.273 / 3.310 / 3.282 | 352 / 348 / 351 |

**Acceptance from branching is +3.5–4% and saturates at two siblings.**
Budgets 12/16/32 buy zero additional steps. From the trace:

- At the first mismatch, the target's token **is** the rank-1 sibling **37%**
  of the time (top-3: 58%). The slots are real.
- But the sibling's *continuation* isn't there: after a sibling is accepted,
  the next depth's marginal is correct only **22%** of the time, vs **65–84%**
  on the spine.

A block-diffusion drafter emits marginals for k positions from one forward
pass, and those marginals are effectively conditioned on the argmax path. So a
sibling is worth exactly one leaf and its children are mostly noise. That's a
property of the drafter family rather than of any implementation, so it may
be worth checking before investing in wider trees. (@liranringel — this is
the part I'd most like a second opinion on, since it bears on which block-wise
drafters the tree would pay off with.)

One caveat if you reproduce this: counterfactuals computed from a single trace
are position-biased — the tree stops exactly where continuation is hard. Ours
read +15% until we cross-checked against end-to-end step counts, which said
+3.4%.

### Three things that may be useful regardless of DDTree

1. **`flashinfer.merge_state` is consistently faster than vLLM's
   `merge_attn_states`** (+0.67 vs +1.00 ms at context 700; +1.30 vs +1.52 at
   3000). Its LSE can be consumed in the wrapper's native layout *and* log2
   domain, which removes a per-layer transpose → contiguous copy → log2-to-ln
   conversion, about 6 kernels per layer. Verified against a log2 reference
   (error 0.0008; natural log gives 0.34). The DCP path has the same shape.
2. **Chain-shaped drafts don't need `custom_mask`** — causal visibility is
   already equivalent. −0.70 ms/step, byte-identical output. Relevant to any
   tree implementation that degenerates to chain shape.
3. **There's a +7 ms/step cliff going from 8 to 9 query tokens on GDN
   hybrids.** It isn't CUDA-graph padding (an exact capture size of 10 changes
   nothing) — the stock fused GDN MTP kernel caps at `state_indices [N,S],
   S ≤ 8`, and above that you fall to a non-fused Triton path: CUDA +3.4 ms
   but CPU +16.5 ms/step, roughly 500 extra launches. A chain never hits it
   because DFlash2's horizon is 8; any tree wider than that does.

Happy to re-run any specific configuration on A6000 if that would help.
